test/wallet-rescan-test.js

```diff
@@ -0,0 +1,516 @@
+    });
+
+    it('should process TRANSFER (again)', async () => {
+      // Alice transfers the name to her own address
+      const aliceTransfer = await alice.createTransfer(NAME, aliceAddr);
+      await node.mempool.addTX(aliceTransfer.toTX());
+      const transferBlocks = await mineBlocks(1);
+      assert.strictEqual(transferBlocks[0].txs.length, 2);
+
+      // Bob detects the TRANSFER even though it doesn't involve him at all
+      const ns = await bob.getNameStateByName(NAME);
+      assert(ns);
+      assert.strictEqual(ns.transfer, node.chain.height);
+
+      // Bob's wallet has not indexed the TRANSFER
+      const bobTransfer = await bob.getTX(aliceTransfer.hash());
+      assert.strictEqual(bobTransfer, null);
+    });
+
+    it('should process REVOKE', async () => {
+      // Alice revokes the name
+      const aliceRevoke = await alice.createRevoke(NAME);
+      await node.mempool.addTX(aliceRevoke.toTX());
+      const revokeBlocks = await mineBlocks(1);
+      assert.strictEqual(revokeBlocks[0].txs.length, 2);
+
+      // Bob detects the REVOKE even though it doesn't involve him at all
+      const ns = await bob.getNameStateByName(NAME);
+      assert(ns);
+      assert.strictEqual(ns.revoked, node.chain.height);
+
+      // Bob's wallet has not indexed the REVOKE
+      const bobTransfer = await bob.getTX(aliceRevoke.hash());
+      assert.strictEqual(bobTransfer, null);
+    });
+
+    it('should fully rescan', async () => {
+      // Complete chain rescan
+      await wdb.rescan(0);
+      await forValue(wdb, 'height', node.chain.height);
+
+      // No change
+      const ns = await bob.getNameStateByName(NAME);
+      assert(ns);
+      assert.strictEqual(ns.revoked, node.chain.height);
+    });
+  });
+
+  describe('Bids, loses, shallow rescan', function() {
+    // Bob runs a full node with wallet plugin
+    const node = new FullNode({
+      network: network.type,
+      memory: true,
+      plugins: [require('../lib/wallet/plugin')]
+    });
+    node.on('error', (err) => {
+      assert(false, err);
+    });
+
+    const {wdb} = node.require('walletdb');
+    let bob, bobAddr;
+
+    // Alice is some other wallet on the network
+    const alice = new MemWallet({ network });
+    const aliceAddr = alice.getAddress();
+
+    // Connect MemWallet to chain as minimally as possible
+    node.chain.on('connect', (entry, block) => {
+      alice.addBlock(entry, block.txs);
+    });
+    alice.getNameStatus = async (nameHash) => {
+      assert(Buffer.isBuffer(nameHash));
+      const height = node.chain.height + 1;
+      const state = await node.chain.getNextState();
+      const hardened = state.hasHardening();
+      return node.chain.db.getNameStatus(nameHash, height, hardened);
+    };
+
+    const NAME = rules.grindName(4, 4, network);
+
+    // Block that confirmed the bids
+    let bidBlockHash;
+    // Hash of the FINALIZE transaction
+    let aliceFinalizeHash;
+
+    async function mineBlocks(n, addr) {
```

Comment:
i feel like all this setup procedure stuff could be a lot more DRY just by moving it up a scope. this could become cumbersome if some of the internal apis change and a lot of test code has to be changed.

Reply:
or if it's special per context it should probably be highlighted what exactly is special in each place

Reply:
true, lots of test modules have a `mineBlocks` function. They are usually slightly different in each case but we could probably add this to a `util` file in a future PR. It would need to accept a chain and miner as parameters though, which are usually local to each test. In this case I need the function to return the blocks, but thats not always the case in other test that mine blocks.

Reply:
gotcha, yeah i could see building up a small library of these sorts of things to raise the level of abstraction for various bob/alice scenarios but you'd need to find the common patterns. obvs all tangential to this branch, just something that sticks out to me as worth improving.